packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,786 @@
+      slot: props.block.message.slot,
+    };
+    const blockInput = new BlockInputColumns(init, state, props.custodyConfig);
+
+    blockInput.blockPromise.resolve(props.block);
+    if (hasAllData) {
+      blockInput.dataPromise.resolve([]);
+    }
+    return blockInput;
+  }
+
+  static createFromColumn(props: AddColumn & CreateBlockInputMeta & {custodyConfig: CustodyConfig}): BlockInputColumns {
+    const hasAllData = props.custodyConfig.sampledColumns.length === 0;
+    const state: BlockInputColumnsState = {
+      hasBlock: false,
+      hasAllData,
+    };
+    const init: BlockInputInit = {
+      daOutOfRange: false,
+      timeCreated: props.seenTimestampSec,
+      forkName: props.forkName,
+      blockRootHex: props.blockRootHex,
+      parentRootHex: toHex(props.columnSidecar.signedBlockHeader.message.parentRoot),
+      slot: props.columnSidecar.signedBlockHeader.message.slot,
+    };
+    const blockInput = new BlockInputColumns(init, state, props.custodyConfig);
+    if (hasAllData) {
+      blockInput.dataPromise.resolve([]);
+    }
+    return blockInput;
+  }
+
+  getLogMeta(): LogMetaColumns {
+    return {
+      blockRoot: prettyBytes(this.blockRootHex),
+      slot: this.slot,
+      expectedColumns:
+        this.state.hasBlock && this.state.block.message.body.blobKzgCommitments.length === 0
+          ? 0
+          : this.custodyConfig.sampledColumns.length,
+      receivedColumns: this.getSampledColumns().length,
+    };
+  }
+
+  addBlock(props: AddBlock<ForkColumnsDA>): void {
+    if (this.state.hasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INVALID_CONSTRUCTION,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot addBlock to BlockInputColumns after it already has a block"
+      );
+    }
+
+    if (props.blockRootHex !== this.blockRootHex) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISMATCHED_ROOT_HEX,
+          blockInputRoot: this.blockRootHex,
+          mismatchedRoot: props.blockRootHex,
+          source: props.source.source,
+          peerId: `${props.source.peerIdStr}`,
+        },
+        "addBlock blockRootHex does not match BlockInput.blockRootHex"
+      );
+    }
+
+    for (const {columnSidecar} of this.columnsCache.values()) {
+      if (!blockAndColumnArePaired(props.block, columnSidecar)) {
+        this.columnsCache.delete(columnSidecar.index);
+        // this.logger?.error(`Removing columnIndex=${columnSidecar.index} from BlockInput`, {}, err);
+      }
+    }
+
+    const hasAllData = props.block.message.body.blobKzgCommitments.length === 0 || this.state.hasAllData;
+
+    this.state = {
+      ...this.state,
+      hasBlock: true,
+      hasAllData,
+      block: props.block,
+      source: props.source,
+      timeCompleteSec: hasAllData ? props.source.seenTimestampSec : undefined,
+    } as BlockInputColumnsState;
+
+    this.blockPromise.resolve(props.block);
```

Comment:
If we resolve `this.state.block` the GC can collect the props object

Reply:
it can be gc'd either way (no ref to `props`)
(not that it buys us much (N pointers to N properties), all properties are used and retained in the block input)